sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -119,4 +121,198 @@ object IntervalUtils {
+      new CalendarInterval(sign * (years * 12 + months), 0)
+    } catch {
+      case e: Exception =>
+        throw new IllegalArgumentException(
+          "Error parsing interval year-month string: " + e.getMessage, e)
+    }
+  }
+
+  /**
+   * Parse dayTime string in form: [-]d HH:mm:ss.nnnnnnnnn and [-]HH:mm:ss.nnnnnnnnn
+   *
+   * adapted from HiveIntervalDayTime.valueOf
+   */
+  def fromDayTimeString(s: String): CalendarInterval = {
+    fromDayTimeString(s, "day", "second")
+  }
+
+  /**
+   * Parse dayTime string in form: [-]d HH:mm:ss.nnnnnnnnn and [-]HH:mm:ss.nnnnnnnnn
+   *
+   * adapted from HiveIntervalDayTime.valueOf.
+   * Below interval conversion patterns are supported:
+   * - DAY TO (HOUR|MINUTE|SECOND)
+   * - HOUR TO (MINUTE|SECOND)
+   * - MINUTE TO SECOND
+   */
+  def fromDayTimeString(input: String, from: String, to: String): CalendarInterval = {
+    if (input == null) {
+      throw new IllegalArgumentException("Interval day-time string was null")
+    }
+    val s = input.trim
+    val m = dayTimePattern.matcher(s)
+    if (!m.matches) {
+      throw new IllegalArgumentException(
+        "Interval string does not match day-time format of 'd h:m:s.n': " + s)
+    }
+    try {
+      val sign = if (m.group(1) != null && m.group(1) == "-") -1
+      else 1
+      val days = if (m.group(2) == null) 0
+      else toLongWithRange("day", m.group(3), 0, Integer.MAX_VALUE)
+      var hours: Long = 0L
+      var minutes: Long = 0L
+      var seconds: Long = 0L
+      if (m.group(5) != null || from == "minute") { // 'HH:mm:ss' or 'mm:ss minute'
+        hours = toLongWithRange("hour", m.group(5), 0, 23)
+        minutes = toLongWithRange("minute", m.group(6), 0, 59)
+        seconds = toLongWithRange("second", m.group(7), 0, 59)
+      }
+      else if (m.group(8) != null) { // 'mm:ss.nn'
+        minutes = toLongWithRange("minute", m.group(6), 0, 59)
+        seconds = toLongWithRange("second", m.group(7), 0, 59)
+      }
+      else { // 'HH:mm'
+        hours = toLongWithRange("hour", m.group(6), 0, 23)
+        minutes = toLongWithRange("second", m.group(7), 0, 59)
+      }
+      // Hive allow nanosecond precision interval
+      val nanoStr = if (m.group(9) == null) null
+      else (m.group(9) + "000000000").substring(0, 9)
+      var nanos = toLongWithRange("nanosecond", nanoStr, 0L, 999999999L)
+      to match {
+        case "hour" =>
+          minutes = 0
+          seconds = 0
+          nanos = 0
+        case "minute" =>
+          seconds = 0
+          nanos = 0
+        case "second" =>
+          // No-op
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Cannot support (interval '$s' $from to $to) expression")
+      }
+      var micros = days * DateTimeUtils.MICROS_PER_DAY
+      micros += hours * MICROS_PER_HOUR
+      micros += minutes * MICROS_PER_MINUTE
+      micros += seconds * DateTimeUtils.MICROS_PER_SECOND
+      micros += nanos / DateTimeUtils.NANOS_PER_MICROS
```

Comment:
This seems to be completely rewritten. Is there any reason for this?

Reply:
because of line length limit and to replace magic constant by `DateTimeUtils.NANOS_PER_MICROS`

Reply:
I made it closer to original code